jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/JibExtension.java

```diff
@@ -118,6 +120,15 @@ public JibExtension(Project project) {
         objectFactory.newInstance(ExtensionParametersSpec.class, pluginExtensions);
     allowInsecureRegistries = objectFactory.property(Boolean.class);
     containerizingMode = objectFactory.property(String.class);
+    configurationName =
+        objectFactory
+            .property(String.class)
+            .convention(
+                project.provider(
+                    () ->
+                        System.getProperty(
+                            PropertyNames.CONFIGURATION_NAME,
+                            JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME)));
```

Comment:
I don't think this works. The system property should take precedence over the value set in `build.gradle`.

Reply:
See my reasoning above, the only workaround possible is probably the one above, but that seems really unidiomatic. I can implement that if you insist, but I'd prefer not to. On the other hand so is a lot of the plugin currently, so that might be something worth considering for a major release of the plugin.

Reply:
We try to think from the user perspective, and I think what is more important than trying to have the code idiomatic is that
1. The user should be able to conveniently override Jib configurations on the command-line (i.e., via system properties).
2. Jib should not build an incorrect image. (When some Jib config was changed on the command-line, Jib should rebuild a new image with the new config.)
I am fine with whatever best and simplest way to achieve these. But what's really important is that these should be satisfied.

Reply:
I'd be in favor of a more standard gradle approach (this is actually the first plugin where I've seen predefined overrides via system properties) that gives to use the ability to opt-in into overrides using his/her technology of choice.

As the case with all gradle options a user may opt in any case to use system properties or environment variables:
```kotlin
jib {
    configurationName.set(System.property("jib.configurationName", "someCustomDefault"))
    // or for example
    configurationName.set(System.getenv("CONFIGURATION_NAME"))
    // or even
    configurationName.set(if (file("abc.txt").exists()) "option1" else "option2")
}
```
This approach is different from what maven is doing I think, but the idiomatic pattern with gradle. Your second point is satisfied right away and the user can opt into the first point by choosing the technology used.

I can see though, why you would't want to change this in a minor new feature (and create inconsistencies in the API).

Reply:
> The user should be able to conveniently override Jib configurations on the command-line

I do understand what you've been saying, but please allow me to re-iterate: being able to conveniently override or set values *on the command-line* without ever touching `build.gradle` or even any file on the target repo is very important. For example, it's even possible, that solely on the command-line, you can run Jib on any Gradle project solely where Jib isn't configured. For example, [this Tekton Jib pipeline task](https://github.com/tektoncd/catalog/blob/8dd769f8e8c6e0f359ee708d5c7f7cb24c885fee/task/jib-gradle/0.2/jib-gradle.yaml#L49-L70) can take any Java source repo as input and containerize the app. The user doesn't even have to know about Jib. 

And there are many users setting system properties only on the command-line (we've seen many cases particularly on CI/CD env), and supporting this is really critical for Jib. (Some examples [here](https://stackoverflow.com/a/64967072/1701388), [here](https://stackoverflow.com/a/61352644/1701388), [here](https://stackoverflow.com/q/59254492/1701388),[here]( https://stackoverflow.com/a/62127210/1701388), and [here]( https://stackoverflow.com/q/57850321/1701388).) And we've learned that, convenience and usability matters a lot for end users.

Reply:
That would still be possible via defaults.

Also all your samples refer to maven and tekton already writes gradle configuration anyway (as said gradle has a different model, so if you need to apply the plugin, you'll have to modify the config anyway, as is the case with any other built-in functionally with gradle.

Reply:
Often, the person building an image is not the person developing an app, and the former person doesn't have control over an app being containerized. (BTW, I don't agree that "Tekton already writes gradle configuration." It's just creates a temp file `init-script.gradle` somewhere (happens to be under `$HOME` but could have been under `/tmp/init-script.gradle`); it doesn't touch the source repo at all.) But even so, Tekton is just one example. For example, Cloud Run Button is a tool that you can add to any public GitHub repos, which then will build the source repo. It prefers using Jib if Jib is already configured on the repo. It first [checks if the Jib plugin is applied](https://github.com/GoogleCloudPlatform/cloud-run-button/blob/ff276c6f2ec7ff1b80710426bdc3d8eaafe3e2d6/cmd/cloudshell_open/jib.go#L45-L46) on the source repo by actually trying to run some Jib command and checking the result, and if the Jib plugin is found on the source repo, it [runs Jib while overriding values](https://github.com/GoogleCloudPlatform/cloud-run-button/blob/ff276c6f2ec7ff1b80710426bdc3d8eaafe3e2d6/cmd/cloudshell_open/jib.go#L27-L28) configured in source repo. Cloud Run Button doesn't yet support Gradle, but Gradle part would work in the exact same manner. If you didn't think Tekton doesn't touch the source repo (which I don't really agree), I am certain you'll agree that this Cloud Run Button for Gradle (if existed) would truly not modify the source repo.

And while these SO examples are not gradle, as I said, from the perspective of a person building an image or working on the CI/CD side or the build platform won't see the difference (as in the Cloud Run Button example). For them, they should just be able to run Jib (whether using Gradle or Maven or anything else; convention doesn't matter) while conclusively overriding parameters already configured in the project. That is, even if the project were Gradle and not Maven, some of the users on the SO (which may not be the dev people but ops) would use the same way to override Jib config values.

> That would still be possible via defaults.

Not sure if I follow this. If you are talking about your last commit change where you set the property convention to a system property value, I believe that doesn't work as we intend. For example, often people define `jib.to.image` in `build.gradle`, which will simply override the default value coming from the system property. But then, the person building an image may want to change the image push destination without touching `builld.gradle` by giving `-Djib.to.image` on the command-line. I see that, with your last commit, the system property will be used if the corresponding value isn't set in `build.gradle`, but you can't override the value in `build.gradle` with `-Djib.to.image` on the command-line, right?

So, with your last commit change, I don't really see an actual migration path for the Tekton or Cloud Run Button examples if Jib drops the current support for system properties.

Reply:
> BTW, I don't agree that "Tekton already writes gradle configuration." It just creates one temp file init-script.gradle somewhere (happens to be under $HOME but could have been under /tmp/init-script.gradle

Well `init-script.gradle` _is_ gradle configuration. Something like that would also be needed for cloud run. This would simply result in something like this (if you want to be sure that you want to override it):
```bash
echo "initscript {
  repositories { maven { url 'https://plugins.gradle.org/m2' } }
  dependencies { classpath 'gradle.plugin.com.google.cloud.tools:jib-gradle-plugin:+' }
}
rootProject {
  afterEvaluate {
    if (!project.plugins.hasPlugin('com.google.cloud.tools.jib')) {
      project.apply plugin: com.google.cloud.tools.jib.gradle.JibPlugin
    }
    jib {
      configurationName = "something"
    }
  }
}" > $HOME/init-script.gradle
```

> I see that, with your last commit, the system property would be used only if the corresponding value isn't set in build.gradle. That is, you can't override the value in build.gradle with -Djib.to.image on the command-line, right?

That is correct and as you really want that specifically, I'll just go an implement it. I'm not sure that discussing this further would bring up any new results.

> So, with your last commit change, I don't really see an actual migration path for the Tekton or Cloud Run Button examples if Jib drops the current support for system properties that can override Jib values set in a project.

See above

Reply:
> See above

Sure, technically it's does not look impossible, but again, it is very important for the user to be able to *conveniently* override Jib config values on the command-line. You will agree that the possible workaround is too cumbersome and brittle. That is simply not acceptable in practice, and I am sure that we will face with a huge backlash. I can bring up more examples (such as Skaffold and Cloud Code for IntelliJ), but seems like I don't need to. It is important for us to keep the current support to be successful.